Samples/NetworkMediaStreamer/RTPMediaStreamer/src/RTPMediaSink.cpp

```diff
@@ -0,0 +1,224 @@
+        return S_OK;
+    }
+
+    STDMETHODIMP GetStreamSinkById(
+        /* [in] */ DWORD dwStreamSinkIdentifier,
+        /* [out] */ __RPC__deref_out_opt IMFStreamSink** ppStreamSink) noexcept override
+    {
+        RETURNIFSHUTDOWN;
+        if (dwStreamSinkIdentifier > m_spStreamSinks.size())
+        {
+            return MF_E_INVALIDSTREAMNUMBER;
+        }
+        else
+        {
+            m_spStreamSinks[dwStreamSinkIdentifier].as<IMFStreamSink>().copy_to(ppStreamSink);
+            return S_OK;
+        }
+    }
+
+    STDMETHODIMP SetPresentationClock(
+        /* [in] */ __RPC__in_opt IMFPresentationClock* pPresentationClock) noexcept override
+    {
+        RETURNIFSHUTDOWN;
+        if (m_spClock)
+        {
+            m_spClock->RemoveClockStateSink(this);
+            m_spClock = nullptr;
+        }
+        m_spClock.copy_from(pPresentationClock);
+        return m_spClock->AddClockStateSink(this);
+
+    }
+
+    STDMETHODIMP GetPresentationClock(
+        /* [out] */ __RPC__deref_out_opt IMFPresentationClock** ppPresentationClock) noexcept override
+    {
+        RETURNIFSHUTDOWN;
+        if (!ppPresentationClock)
+        {
+            return E_POINTER;
+        }
+        m_spClock.copy_to(ppPresentationClock);
+        return S_OK;
+    }
+
+    STDMETHODIMP Shutdown(void) noexcept
+    {
+        RETURNIFSHUTDOWN;
+        HRESULT hr = S_OK;
+        m_bIsShutdown = true;
+        for (auto strm : m_spStreamSinks)
+        {
+            auto hr1 = strm->Shutdown();
+            hr = FAILED(hr1) ? hr1 : hr;
+        }
+        return hr;
+    }
+
+    //IMediaExtension
+    STDMETHODIMP SetProperties(
+        winrt::Windows::Foundation::Collections::IPropertySet const& configuration) noexcept
+    {
+        return S_OK;
+    }
+
+    // IMFClockStateSink methods
+    STDMETHODIMP OnClockStart(MFTIME hnsSystemTime, LONGLONG llClockStartOffset) noexcept override
+    {
+        HRESULT hr = S_OK;
+        for (auto strm : m_spStreamSinks)
+        {
+            auto hr1 = strm->Start(hnsSystemTime, llClockStartOffset);
+            hr = FAILED(hr1) ? hr1 : hr;
+        }
+        return hr;
+    }
+
+    STDMETHODIMP OnClockStop(MFTIME hnsSystemTime) noexcept override
+    {
+        HRESULT hr = S_OK;
+        for (auto strm : m_spStreamSinks)
+        {
+            auto hr1 = strm->Stop(hnsSystemTime);
+            hr = FAILED(hr1) ? hr1 : hr;
+        }
+        return hr;
+    }
+
+    STDMETHODIMP OnClockPause(MFTIME hnsSystemTime) noexcept override
+    {
+        HRESULT hr = S_OK;
+        for (auto strm : m_spStreamSinks)
+        {
+            auto hr1 = strm->Pause(hnsSystemTime);
+            hr = FAILED(hr1) ? hr1 : hr;
+        }
+        return hr;
+    }
+
+    STDMETHODIMP OnClockRestart(MFTIME hnsSystemTime) noexcept override
+    {
+        HRESULT hr = S_OK;
+        for (auto strm : m_spStreamSinks)
+        {
+            auto hr1 = strm->Start(hnsSystemTime, 0);

```

Comment:
why continue start rest of streams if one of them failed?

Reply:
This never retruns failure in present situation. But this logic is put in  to represent a high availablity system, where any available stream should work. RTP handles both streams separately.

---
In reply to: [494111042](https://github.com/microsoft/Windows-Camera/pull/13#discussion_r494111042) [](ancestors = 494111042)

Reply:
just curious, if it does how the caller usually handle this?

Reply:
need to check SinkWriter implementation for that..

---
In reply to: [494524524](https://github.com/microsoft/Windows-Camera/pull/13#discussion_r494524524) [](ancestors = 494524524)

Reply:
it seems very hazy on how the called should treat the error. But most multi-stream sinks in the OS are doing the same logic as here. Some even ignore the error.

---
In reply to: [494625563](https://github.com/microsoft/Windows-Camera/pull/13#discussion_r494625563) [](ancestors = 494625563,494524524)